File: src/rest/import.ts
import type { OpenAPIHono } from "@hono/zod-openapi";
import type { AuthEnv } from "../auth/middleware.js";
import { ValidationError } from "../core/errors.js";
import type { Services } from "../services/container.js";

/** Multipart .ics import. Plain Hono route (not OpenAPI-documented). */
export function registerImportRoutes(
	api: OpenAPIHono<AuthEnv>,
	{ importer, calendars }: Services,
): void {
	api.post("/events/import", async (c) => {
		const body = await c.req.parseBody();
		const file = body.file;
		const calendarId = body.calendar_id;
		if (!(file instanceof File))
			throw new ValidationError("Expected a 'file' field");
		if (typeof calendarId !== "string" || !calendarId)
			throw new ValidationError("Expected a 'calendar_id' field");
		calendars.get(calendarId); // throws NotFoundError (404) for an unknown id
		if (file.size > 10 * 1024 * 1024)
			throw new ValidationError("File too large (max 10 MB)");
		const text = await file.text();
		return c.json(importer.importIcs(text, calendarId), 200);
	});
}

File: src/rest/import.rs
use axum::{Router, Json, extract::State};
use axum::routing::post;
use axum::http::StatusCode;
use crate::core::error::AppError;
use crate::rest::AppState;

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/events/import", post(import_ics))
}

async fn import_ics(
    State(state): State<AppState>,
    mut multipart: axum::extract::Multipart,
) -> Result<(StatusCode, Json<serde_json::Value>), AppError> {
    let mut file_data: Option<Vec<u8>> = None;
    let mut calendar_id: Option<String> = None;

    while let Some(field) = multipart.next_field().await.map_err(|_| AppError::validation("Invalid multipart", None))? {
        match field.name() {
            Some("file") => {
                file_data = Some(
                    field.bytes().await
                        .map_err(|_| AppError::validation("Failed to read file", None))?
                        .to_vec()
                );
            }
            Some("calendar_id") => {
                calendar_id = Some(
                    field.text().await
                        .map_err(|_| AppError::validation("Failed to read calendar_id", None))?
                );
            }
            _ => {}
        }
    }

    let data = file_data.ok_or_else(|| AppError::validation("Missing file field", None))?;
    let cal_id = calendar_id.ok_or_else(|| AppError::validation("Missing calendar_id field", None))?;

    let result = state.db.call(move |conn| {
        crate::services::import::import_ics(conn, &cal_id, &data)
    }).await?;

    // One coarse broadcast per import: listeners refetch the whole event list,
    // so per-event emits would only flood the bus on large files.
    if result.created > 0 {
        state.bus.emit("event.created", serde_json::json!({"imported": result.created}));
    }

    Ok((StatusCode::OK, Json(serde_json::to_value(result).unwrap())))
}
